optimize PFDisplacedVertexCandidateFinder::link

jpata opened the pull request:
#### PR description:

Technical optimization of `PFDisplacedVertexCandidateFinder::link` as suggested in https://github.com/cms-sw/cmssw/issues/30955. Cache pt, eta, phi values using SOA, rearrange the calculations to do the crossing-point evaluation only when needed. A further refactoring of this class could be done (fabs, revisit recursive associate algo) but was not addressed so far.

#### PR validation:

workflow `23434.21` has been run with no change in output.

Initial checks show about a 25% speed improvement for `PFDisplacedVertexCandidateFinder::link` and `PFDisplacedVertexCandidateProducer::produce` correspondingly.

before:
```
[149]       1.5     128.48      16.64 / 111.84     PFDisplacedVertexCandidateFinder::link(edm::RefToBase<reco::Track> const&, edm::RefToBase<reco::Track> const&, double&, Point3DBase<float, GlobalTag>&, reco::PFDisplacedVertexCandidate::VertexLinkTest&)
            1.0  .........      87.30 / 97.57        TwoTrackMinimumDistance::calculate(GlobalTrajectoryParameters const&, GlobalTrajectoryParameters const&) [185]
            0.3  .........      21.68 / 132.91       __ieee754_log_avx [144]
            0.0  .........       1.67 / 2.96         edm::refitem::GetRefPtrImpl<std::vector<reco::Track, std::allocator<reco::Track> >, reco::Track, edm::refhelper::FindUsingAdvance<std::vector<reco::Track, std::allocator<reco::Track> >, reco::Track>, unsigned int>::getRefPtr_(edm::RefCore const&, unsigned int) [2203]
            0.0  .........       0.61 / 3.16         log [2161]
            0.0  .........       0.20 / 0.21         edm::reftobase::Holder<reco::Track, edm::Ref<std::vector<reco::Track, std::allocator<reco::Track> >, reco::Track, edm::refhelper::FindUsingAdvance<std::vector<reco::Track, std::allocator<reco::Track> >, reco::Track> > >::getPtr() const [5773]
            0.0  .........       0.10 / 0.12         TwoTrackMinimumDistance::distance() const [6931]
            0.0  .........       0.09 / 97.25        __ieee754_atan2_avx [187]
            0.0  .........       0.08 / 0.08         TwoTrackMinimumDistance::crossingPoint() const [7761]
            0.0  .........       0.05 / 10.34        edm::reftobase::Holder<reco::Track, edm::Ref<std::vector<reco::Track, std::allocator<reco::Track> >, reco::Track, edm::refhelper::FindUsingAdvance<std::vector<reco::Track, std::allocator<reco::Track> >, reco::Track> > >::key() const [1493]
            0.0  .........       0.04 / 49.53        _init [306]
```

after:
```
[186]       1.1      96.53       3.82 / 92.71      PFDisplacedVertexCandidateFinder::link(edm::RefToBase<reco::Track> const&, edm::RefToBase<reco::Track> const&, double&, Point3DBase<float, GlobalTag>&, reco::PFDisplacedVertexCandidate::VertexLinkTest&)
            1.1  .........      92.39 / 102.49       TwoTrackMinimumDistance::calculate(GlobalTrajectoryParameters const&, GlobalTrajectoryParameters const&) [178]
            0.0  .........       0.17 / 0.18         TwoTrackMinimumDistance::distance() const [6143]
            0.0  .........       0.07 / 10.12        edm::reftobase::Holder<reco::Track, edm::Ref<std::vector<reco::Track, std::allocator<reco::Track> >, reco::Track, edm::refhelper::FindUsingAdvance<std::vector<reco::Track, std::allocator<reco::Track> >, reco::Track> > >::key() const [1504]
            0.0  .........       0.04 / 0.06         TwoTrackMinimumDistance::crossingPoint() const [8743]
            0.0  .........       0.03 / 51.14        _init [296]
```

The majority of time of `PFDisplacedVertexCandidateFinder::link` is now spent in `TwoTrackMinimumDistance::calculate` if I interpret the igprof output correctly. I think it could be useful to understand if there are algorithmic improvements available.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

A backport is not planned at this moment.

jpata committed:
optimize PFDisplacedVertexCandidateFinder::link

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17706

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CommonTools/Utils/interface/KinematicTables.h  modified in PR(s): #30956


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17706/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17706/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

jpata committed:
code check

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17707

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CommonTools/Utils/interface/KinematicTables.h  modified in PR(s): #30956

cmsbuild commented:
A new Pull Request was created by  @jpata (Joosep Pata) for master.

It involves the following packages:

CommonTools/Utils
RecoParticleFlow/PFTracking


@perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@mmarionncern, @makortel, @cbernet, @rovere, @lgray, @lecriste, @hatakeyamak, @seemasharmafnal this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

jpata commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: a13641830a120ba8605c25c4e11d93722ac03dab
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8721/summary.html
CMSSW: CMSSW_11_2_X_2020-08-11-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8721/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612401
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2612347
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jpata committed:
revert crossingPoint

jpata commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on CommonTools/Utils/interface/KinematicTables.h:
```diff
@@ -22,6 +22,16 @@ namespace edm {
+    }
+
     template <class Object>
     auto makeEtaPhiTableLazy(std::vector<Object> const& objects) {
```
this is kind of left dangling; it would be better to have a more continuous implementation: etaPhi first, then ptEtaPhi.

jpata replied:
pinging @guitargeek as the author of `KinematicTables` - any suggestions as to how I could improve this would be much appreciated!

guitargeek replied:
Hi Joosep, I think Slavas comment refers just to the order of the statements in this file. It's a bit confusing that there is there first the `makeEtaPhiTable` function, then the `makePtEtaPhiTable`, and then it goes back to only EtaPhi with `makeEtaPhiTableLazy`.

Other than this possible reordering, I would not bother changing anything else in this file in this PR, in particular not anything related to the column fillers. This is under development already in https://github.com/cms-sw/cmssw/pull/30956. We can just rebase whichever PR gets merged second (it will probably be mine). Is that okay for you?

slava77 replied:
> Hi Joosep, I think Slavas comment refers just to the order of the statements in this file, right? It's a bit confusing that there is there first the `makeEtaPhiTable` function, then the `makePtEtaPhiTable`, and then it goes back to only EtaPhi with `makeEtaPhiTableLazy`.

right, that's what I meant

jpata replied:
Oh, got it, thanks. I thought the idea is to reduce the duplication a bit (don't have a good idea for it immediately).

slava77 replied:
looking at https://github.com/cms-sw/cmssw/blob/master/FWCore/SOA/interface/Table.h

```using PtEtaPhiTable = AddColumns_t<EtaPhiTable, col::Pt>;```

appears to be the direction to go with reusing the available EtaPhi.
I'm not sure about the related methods if there is a transparent way to extend

guitargeek replied:
Very soon these functions won't be necessary anyway. When de default fillers are implemented with #30956, the only justification for their existence is to support the lazy construction via `LazyResult`. While this `LazyResult` wrapper is nice, I have now realized that it can't be conveniently used to construct objects without the `makeEtaPhiTableLazy` wrapper function in this file.

We should maybe consider implementing a `LazyConstructed` wrapper to simplify this (maybe in my PR once I solved the other problems). Then, in the producers one would use the table as:
```C++
auto etaPhiTable = LazyConstructed<edm::soa::EtaPhiTable>(tracks);
```
This `KinematicTables.h` file would then only have the `using` statements, where the tables build on top of each other with `AddColumns_t` as Slava said.

slava77 commented on RecoParticleFlow/PFTracking/src/PFDisplacedVertexCandidateFinder.cc:
```diff
@@ -225,6 +235,8 @@ void PFDisplacedVertexCandidateFinder::link(const TrackBaseRef& el1,
     return;
   }
 
+  P = theMinimum_.crossingPoint();
```
since this code is modified, perhaps update the name to stay with CMSSW naming style (no single-letter, no first-capital for locals)

jpata replied:
addressed in https://github.com/cms-sw/cmssw/pull/31121/commits/965eff78a97350337cf4e2f8cb640153834a8b10

slava77 commented on RecoParticleFlow/PFTracking/src/PFDisplacedVertexCandidateFinder.cc:
```diff
@@ -198,24 +199,33 @@ void PFDisplacedVertexCandidateFinder::link(const TrackBaseRef& el1,
                                             double& dist,
                                             GlobalPoint& P,
                                             PFDisplacedVertexCandidate::VertexLinkTest& vertexLinkTest) {
```
this is called from PFDisplacedVertexCandidateFinder::associate, which is itself called recursively; after that this method is also called in ::packLinks.
It's not obvious to me that the same el1,el2 link is tested only once.
If there are repeated calls there is some opportunity for optimization

jpata replied:
yes - I had that in mind with "algorithmic improvements". checking it now...

jpata replied:
I think there is a CPU-memory tradeoff here. `associate` calls `link` just once for each pair, but in order to run `packLinks`, you need to know how many elements were found in total. Therefore, one either needs to recompute the link at packing time (current approach) or cache `dist`, `GlobalPoint` and `linktest` for each pair until the end. Caching could be viable, but depends on the number of elements in a displaced candidate. I checked, and most have a few only, however candidates with thousands of tracks also exist (similar thing happens with linking if PFBlocks).

Therefore, I can't think of a way of improving this easily at the moment, without re-engineering the algorithm with potential physics changes. Unless I missed something simple...

slava77 replied:
I'm guessing that O(10 MB) in-produce (something that goes away after ::produce exits) should be acceptable.  So, if it's thousands of track pairs then we are probably still OK, if it's thousands^2, it may be too much.

jpata replied:
The case is that `packLinks` is doing all pairs of tracks in the displaced candidate to build the distance matrix [link](https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFTracking/src/PFDisplacedVertexCandidateFinder.cc#L285) `associate` only considers a small number of tracks to build the candidate based on a linked list.

Therefore, caching `link` outputs won't be helpful here to save CPU time.

The bigger issue is that an event always seems to contain one pathological displaced candidate with a few thousand tracks. I'm not sure if this candidate results in anything useful down the line. Overall, I think some algorithmic improvements should be considered to avoid building these huge candidates, but this might require a closer look that physics performance is not reduced.

@bendavid @hatakeyamak what are your thoughts?

slava77 reviewed (commented):
(if the browsable report is not working) for completeness, please add the cost of PFDisplacedVertexCandidateFinder::setInput

cmsbuild commented:
+1
Tested at: a13641830a120ba8605c25c4e11d93722ac03dab
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8724/summary.html
CMSSW: CMSSW_11_2_X_2020-08-12-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8724/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612401
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2612352
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17738

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CommonTools/Utils/interface/KinematicTables.h  modified in PR(s): #30956

cmsbuild commented:
Pull request #31121 was updated. @perrotta, @jpata, @santocch, @slava77 can you please check and sign again.

jpata commented:
I reran the igprof profiles just to be sure, and we can shave off about 15% in total from `PFDisplacedVertexCandidateProducer::produce`, which includes both `setInput` and `findDisplacedVertexCandidates`.

profile links: [new](https://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/31121/new/205), [original](https://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/31121/orig/173)

jpata committed:
rename variable

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17739

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CommonTools/Utils/interface/KinematicTables.h  modified in PR(s): #30956

cmsbuild commented:
Pull request #31121 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 965eff78a97350337cf4e2f8cb640153834a8b10
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8743/summary.html
CMSSW: CMSSW_11_2_X_2020-08-13-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8743/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612401
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2612347
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jpata committed:
reorganize

jpata committed:
refactor

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17755

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CommonTools/Utils/interface/KinematicTables.h  modified in PR(s): #30956


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17755/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17755/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

jpata commented:
I've addressed the comments by @slava77 (thanks!)

- the current `associate` algo results in a `PFDisplacedVertexCandidate` that contains `N~10^3` of tracks with `N(N-1)` link calculations being necessary in `packLinks`. I'm not sure if such high-multiplicity displaced vertex candidates are reasonable, but this feature has also been seen in PFBlockAlgo, where it's likely not desireable.
- nevertheless, there are about 15% duplicate link calculations which could be alleviated with caching, but it would require to store a couple of floats for a few hundred thousand track pairs (wf `23434.21`)
- therefore, a decision has to be made if it's worth it to trade off CPU to RSS here, or rather possibly re-study the displaced vertex algo a bit. I could imagine to cut adding tracks to a candidate if it has too many, but the impact of any change needs to be studied.

Minor change:
-  switched `cout` to `edm::LogDebug` as was done in PFAlgo.cc, example log file http://jpata.web.cern.ch/jpata/reco/31121/log.txt (~270MB)

jpata committed:
use std abs

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17756

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CommonTools/Utils/interface/KinematicTables.h  modified in PR(s): #30956

cmsbuild commented:
Pull request #31121 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 568bf3df81181ac34056a6ae6fc5e6906079f36a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8761/summary.html
CMSSW: CMSSW_11_2_X_2020-08-13-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8761/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2608246
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2608192
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jpata commented:
This PR is ready for review from my point of view. There are no reco differences, I wrote the summary of changes here: https://github.com/cms-sw/cmssw/pull/31121#issuecomment-674066156 and the profiler output here: https://github.com/cms-sw/cmssw/pull/31121#issuecomment-673550980.

jpata commented:
@hatakeyamak @bendavid would you have any comments on this techical/refactoring/optimization PR? After your inputs, I will move forward with it.

santocch commented:
+1

hatakeyamak commented:
> @hatakeyamak @bendavid would you have any comments on this techical/refactoring/optimization PR? After your inputs, I will move forward with it.

Looks good to me. el_table_ could possibly be named a little more clearly to show what table this is?

jpata committed:
rename variable

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31121/17813

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CommonTools/Utils/interface/KinematicTables.h  modified in PR(s): #30956

cmsbuild commented:
Pull request #31121 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5911bd4609ce32d3cef9014636c4e7e3488c9c58
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8822/summary.html
CMSSW: CMSSW_11_2_X_2020-08-19-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-89e557/8822/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2608222
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2608194
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jpata commented:
+1

- technical, tests show no relevant reco difference
- time reduced by about 25% by using SOA for caching repeated eta, phi calls
- also refactor to use fabs->std::abs, edm logging, better variable names
- addresses issue https://github.com/cms-sw/cmssw/issues/30955

qliphy commented:
+1
